add numpy 2 hotfix to main

JamesRobertsonGames opened the pull request:
This PR introduces new functionality for checking and updating package dependencies for numpy 2.0 compatibility. It also includes significant updates to the documentation to reflect these changes and provide guidance on using the new features.
### Links

- [PKG-4837](https://anaconda.atlassian.net/browse/PKG-4837) 

### Explanation of changes:
- Added new script `numpy2.py` for checking and proposing numpy 2.0 compatibility updates
- Modified `main.py` to incorporate changes from `proposed_numpy_changes.json`
- Updated README with instructions for using `numpy2.py` and the updated `main.py`
- Added functionality to generate CSV files for reviewing proposed changes
- Included guidance in README for reviewing CSV updates using various tools
- Refactored existing code for better readability and compliance with flake8 standards
- Added error handling and logging improvements in the scripts
- Updated documentation to reflect the new workflow involving `numpy2.py` and `main.py`

[PKG-4837]: https://anaconda.atlassian.net/browse/PKG-4837?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Note to Reviewers:**

This implementation has been carefully designed. However, I value your opinions and welcome any feedback on the design or implementation. This is the best attempt currently at a balanced approach considering future hotfix applications will need to be done carefully. 

Also the packages in the "protect_dict" are just placeholder for now, once this is approved as an approach there will be work to refine this list

cbouss commented on main.py:
```diff
@@ -261,6 +261,42 @@
+                depends[i] = "numpy <2.0a0"
+                updated = True
+            # Add an upper bound if it is missing in the dependency constraint
+            elif '==' not in dep and '<' not in dep:
+                depends[i] = dep + ",<2.0a0"
+                updated = True
```
I see in the output lines like "numpy-base 1.11.3 py27hdbf6ddf_8", replaced by "numpy-base 1.11.3 py27hdbf6ddf_8,<2.0a0",. The upper bound is added on the hash, which I expect to cause trouble, but most of all, we don't need to add upper bounds to packages pinned to a specific version.

cbouss commented on main.py:
```diff
@@ -261,6 +261,42 @@
+    depends = record.get("depends", [])
+    updated = False
+    for i, dep in enumerate(depends):
+        if dep.startswith("numpy"):
```
```suggestion
        if dep.split()[0] == "numpy":
```
Otherwise it will also match numpy-base, numpydoc...

cbouss commented on main.py:
```diff
@@ -261,6 +261,42 @@
+    depends = record.get("depends", [])
+    updated = False
+    for i, dep in enumerate(depends):
+        if dep.startswith("numpy-base") and '==' not in dep and '<' not in dep:
```
```suggestion
        if dep.split()[0] == "numpy-base" and '==' not in dep and '<' not in dep:
```

cbouss commented on main.py:
```diff
@@ -261,6 +261,42 @@
+        if dep.startswith("numpy-base") and '==' not in dep and '<' not in dep:
+            depends[i] = dep + ",<2.0a0"
+            updated = True
+        elif dep == "numpy-base":
```
```suggestion
        else:
```

JamesRobertsonGames commented:
@cbouss thanks for the review, made the suggestions and reuploaded the most upto date output

cbouss commented on main.py:
```diff
@@ -261,6 +261,38 @@
+        if dep.split()[0] == "numpy-base" and '==' not in dep and "1." not in dep and '<' not in dep:
+            depends[i] = dep + ",<2.0a0"
+            updated = True
+        else:
+            depends[i] = "numpy-base <2.0a0"
+            updated = True
```
This causes all dependencies to be replaced by numpy-base

JamesRobertsonGames commented:
Much better output now in the output.txt @cbouss

cbouss commented:
Besides the diff it would be great to have some analysis on the repodata before and after the patch. I feel this would help ensure all is covered.

Also, I have been thinking, what if we publish this patch, then start building python packages with no upper bounds (because they support numpy 2), and someone regenerate a patch? A <1 bound would be wrongfully added. So perhaps instead we could have a configuration file like:
```
pandas: 2.2.2,
scikit-lear: 1.4.2,
...
```
Which could be used to patch like (pseudo code):
```
if pkg in numpy2_protect_dict:
  if VersionOrder(version) <= VersionOrder(numpy2_protect_dict['version']):
    add_dep(numpy < 2.0a0)
```
This way if needed we can tweak the configuration file and retain more control.

JamesRobertsonGames commented on main.py:
```diff
@@ -260,6 +261,60 @@
     "_anaconda_depends",
 ]
 
+# Load the configuration file
+with open('numpy2_protect.yaml', 'r') as f:
+    numpy2_protect_dict = yaml.safe_load(f)
```
My only concern is it is likely to be fairly long? I could add it in a new python file (which just has a giant dict) and include that? This `main.py` is already bloated and long I don't really want to add so 100's of lines potentially (it's 3 at the moment but we haven't compiled the final list)

Jrice1317 replied:
I agree adding pyyaml does seem a bit like overkill for this single usage. Perhaps since `json` is already being imported, that can be utilized here?

JamesRobertsonGames replied:
`json` wasn't selected just because it's slightly less human readable, might as well go forward with a dict in a separate file I think that strikes the balance, and if I do end up writing a script to generate it I can still do that with a separate .py file

cbouss commented on README.md:
```diff
@@ -45,12 +45,76 @@ We should put things on the revoke list when:
+1. Scans through the repodata for packages depending on numpy.
+2. Checks if these dependencies need updates to ensure compatibility with numpy 2.0.
+3. Proposes changes to add upper bounds to numpy dependencies where necessary.
+4. Generates a `proposed_numpy_changes.json` file containing all proposed changes.
```
Can you rename `proposed_numpy_changes.json` to `numpy2_patch.json`? Since this file is checked in and will remain this will be clearer.

cbouss commented on main.py:
```diff
@@ -261,6 +275,134 @@
 
+def load_numpy_changes():
+    try:
+        with open('proposed_numpy_changes.json', 'r') as f:
```
Not having this file should fail the script.

cbouss commented on main.py:
```diff
@@ -460,6 +602,20 @@ def _fix_cudnn_depends(depends, subdir):
     depends[idx] = correct_cudnn_depends
 
 
+def filter_packages(repo_data, substring):
+
+    # Initialize the new dictionary to store filtered results
+    filtered_dict = {}
+
+    # Loop through each key-value pair in the 'Packages' dictionary
+    for key, value in repo_data.items():
+        # If the key contains the desired substring, add it to the filtered dictionary
+        if substring in key:
+            filtered_dict[key] = value
+
+    return filtered_dict
+
+
```
```suggestion
```
I don't see that function being used.

cbouss commented on main.py:
```diff
@@ -261,6 +275,134 @@
+        return {}
+
+
+NUMPY_CHANGES = load_numpy_changes()
+
+
+def apply_numpy_changes(record, subdir, filename):
```
Can you make it more explicit? Perhaps just replacing numpy by numpy2.

cbouss commented on numpy2.py:
```diff
@@ -0,0 +1,255 @@
+CHANNEL_NAME = "main"
+CHANNEL_ALIAS = "https://repo.anaconda.com/pkgs"
+SUBDIRS = (
+    "noarch",
+    "linux-32",
+    "linux-64",
+    "linux-aarch64",
+    "linux-armv6l",
+    "linux-armv7l",
+    "linux-ppc64le",
+    "linux-s390x",
+    "osx-64",
+    "osx-arm64",
+    "win-32",
+    "win-64",
```
```suggestion
    "noarch",
    "linux-64",
    "linux-aarch64",
    "linux-s390x",
    "osx-64",
    "osx-arm64",
    "win-64",
```
Only need to work on what we support today.

cbouss commented on numpy2.py:
```diff
@@ -0,0 +1,255 @@
+            constrains = record.get("constrains", [])
+
+            depends = [dep for dep in depends if dep is not None]
+            if "py39" in fn or "py310" in fn or "py311" in fn or "py312" in fn:
```
```suggestion
            # numpy 2 is introduced for python 3.9. Packages for python 3.13 will be built with numpy2 from the start.
            if "py39" in fn or "py310" in fn or "py311" in fn or "py312" in fn:
```

cbouss replied:
This leaves noarch packages out though. Perhaps that's fine.

cbouss reviewed (commented):
Did a quick first pass

cbouss commented on numpy2_patch.json:
Can you generate it avoiding the case where no update actually occurs?
I mean removing entries like:
```
    "adtk-0.6.2-py310h2f386ee_0.tar.bz2": [
      {
        "type": "dep",
        "original": "numpy >=1.21.5,<2.0a0",
        "updated": "numpy >=1.21.5,<2.0a0",
        "reason": "Already has upper bound"
      }
    ],
```

JamesRobertsonGames replied:
Done and regenerated the file 👍

ryanskeith commented on main.py:
```diff
@@ -273,6 +284,95 @@
+    """
+    if subdir not in NUMPY_2_CHANGES or filename not in NUMPY_2_CHANGES[subdir]:
+        return
+    changes = NUMPY_2_CHANGES[subdir][filename]
```
You can then refer to subdir_changes from above.

ryanskeith commented on main.py:
```diff
@@ -273,6 +284,95 @@
+    return None
+
+
+def _apply_changes_to_dependencies(depends, change, record, filename, sort_type='reason'):
```
```suggestion
def _apply_changes_to_dependencies(depends_or_constraints, change, record, filename, sort_type='reason'):
```

ryanskeith reviewed (changes requested):
After thinking on this over the weekend, I feel there are two major things to ask for (if I understand the code correctly).

First, let's simplify numpy2.py code.
- Let's rename to generate_numpy2.py changes
- Remove all code not being used and most reporting type code.
- Simplify the data structure in the numpy2_patch.json
- Move testing code to the tests folder

Second, let's simplify the incorporation of the numpy2 patches into main.py
- Remove any logging or side product reporting
- Use standard dep replacement code

JamesRobertsonGames commented on numpy2.py:
```diff
@@ -0,0 +1,239 @@
+    for i, dependency in enumerate(dependencies_list):
+        parts = dependency.split()
+        package_name = parts[0]
+        if package_name in ["numpy", "numpy-base"]:
```
As discussed: this is correct right now as we have not properly populated the protected versions

JamesRobertsonGames commented:
@cbouss @ryanskeith 

Should have addressed the comments, the output is looking good on the test-hotfix.py to me, removed all non-essential logging and simplified the solution based on the suggestions. Let me know if I have missed anything or anything new is observed :)

ryanskeith commented on main.py:
```diff
@@ -273,6 +273,51 @@
+        replace_dep(depends, change["original"], change["updated"])
+
+
+def _get_dependency_list(record, change_type):
```
I don't think this warrants its own function.  If you store "depends" and "constrains" directly in the numpy2_patch.json file, the replace_dep line becomes:

```
replace_dep(record[change["change_type"]], change["original"], change["updated"]
```

Your process already did all the cleaning so we shouldn't have any weird third cases.

ryanskeith replied:
This also gets rid of the unnecessary for loop.

ryanskeith commented on main.py:
```diff
@@ -686,6 +731,9 @@ def patch_record_in_place(fn, record, subdir):
                 depends[i] = depends[i].replace(">=1.21.5,", ">=1.21.2,")
                 break
 
+    if NUMPY_2_CHANGES:
```
As stated before, as this block changes many different files and NOT a targeted file.  It should be run last.

ryanskeith commented on generate_numpy2_patch.py:
```diff
@@ -0,0 +1,211 @@
+                    version_str = parts[1] if len(parts) > 1 else None
+                    version = parse_version(version_str) if version_str else None
+                    protected_version = parse_version(numpy2_protect_dict[package_name])
+                    if version and protected_version:
```
ah, I see.

ryanskeith commented on main.py:
```diff
@@ -273,6 +273,30 @@
+    if not change:
+        return
+    else:
+        try:
```
In what case do you feel that you will have a failure when you have been given a filename the record will fail?  I don't think you need to be so protective here.

ryanskeith commented on generate_numpy2_patch.py:
```diff
@@ -0,0 +1,213 @@
+                    try:
+                        for dep in depends:
+                            if dep.split()[0] in ["numpy", "numpy-base"]:
+                                update_numpy_dependencies(depends, record, "dep", subdir, fn)
```
```suggestion
                                update_numpy_dependencies(depends, record, "depends", subdir, fn)
```

ryanskeith commented on generate_numpy2_patch.py:
```diff
@@ -0,0 +1,213 @@
+                                update_numpy_dependencies(depends, record, "dep", subdir, fn)
+                        for constrain in constrains:
+                            if constrain.split()[0] in ["numpy", "numpy-base"]:
+                                update_numpy_dependencies(constrains, record, "constr", subdir, fn)
```
```suggestion
                                update_numpy_dependencies(constrains, record, "constrains", subdir, fn)
```

ryanskeith commented on generate_numpy2_patch.py:
```diff
@@ -0,0 +1,213 @@
+    - reason: The reason for the change.
+    """
+    # change dep and constr to dependency and constraint
+    if change_type == "dep":
```
By giving the full change_type on creation, you don't need this if else statement.

ryanskeith replied:
You can get rid of lines 55-58 now.

ryanskeith commented on main.py:
```diff
@@ -1444,6 +1466,14 @@ def patch_record_in_place(fn, record, subdir):
+    # numpy 2 #
+    ###########
+
+    # Numpy 2.0.0 Patch
```
Needs comments on what this does.

ryanskeith commented on main.py:
```diff
@@ -273,6 +273,30 @@
+    change = NUMPY_2_CHANGES[subdir].get(filename)
+    if not change:
+        return
+    else:
```
We don't need an else in this case.

ryanskeith commented on generate_numpy2_patch.py:
```diff
@@ -0,0 +1,213 @@
+                        logger.error(f"numpy 2.0.0 error {fn}: {e}")
+
+    json_filename = Path("numpy2_patch.json")
+    with json_filename.open('w') as f:
```
Not necessary but this can be a one liner.
```suggestion
    json_filename.write_text(json.dumps(dict(NUMPY_2_CHANGES), indent=2)
```

ryanskeith reviewed (changes requested):
This is progressing quite well.  A couple things.

- I think we can safely drop "numpy-base" from the list as it has never been triggered
  - This should simplify the logic in places so you can directly ask and not have to do the reverse.
- I think as this code has a lot of nested if/else statements, you need more comments inline as the logic is hard to follow.  My gut tells me there is still some simplifications that should occur.
  - Specifically more comments for generate_numpy2_patch:update_numpy_dependencies and :main
- Lastly, how are you checking for no arch again?  the filename will not work in the noarch case.

JamesRobertsonGames commented:
@ryanskeith done your suggestions, on the noarch I think a previous review concluded that there is little point doing noarch in this instance? @cbouss if you have an opinion on this feel free to wade in! Happy to handle this edge case, or we could deploy without this and then rectify the noarch packages in another PR?

cbouss commented:
> @ryanskeith done your suggestions, on the noarch I think a previous review concluded that there is little point doing noarch in this instance? @cbouss if you have an opinion on this feel free to wade in! Happy to handle this edge case, or we could deploy without this and then rectify the noarch packages in another PR?

Yes I believe it is best to not patch noarch packages, at least for now. I am not sure of which edge cases this would uncover, and it is now time to release numpy 2.

cbouss reviewed (approved):
LGTM!

ryanskeith reviewed (approved):
Looks good.  You could remove the dead code in `generate_numpy2_patch.py`, but it isn't necessary.
